src/vidpf.rs

```diff
@@ -0,0 +1,667 @@
+        VdafError,
+    },
+};
+
+/// Creates a VIDPF instance where the function F: {0.1}* --> [Range].
+#[must_use]
+pub const fn new<R>(bind_info: Vec<u8>) -> Instance<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    Instance {
+        bind_info,
+        _cd: PhantomData,
+    }
+}
+
+/// Contains the global parameters of a VIDPF instance.
+///
+/// The function is defined as F: {0.1}* --> [Range].
+pub struct Instance<R>
+where
+    // Represents the codomain of a Distributed Point Function (DPF).
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    // Used to cryptographically bind some information.
+    bind_info: Vec<u8>,
+
+    _cd: PhantomData<R>,
+}
+
+impl<R> Instance<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    /// The gen method splits the point function `F(alpha) = beta` in two private keys
+    /// used by the aggregation servers, and a common public key.
+    /// The `beta` value must be a non-zero value.
+    ///
+    /// # Errors
+    /// Triggers an error if the random generator fails, or when beta is zero.
+    pub fn gen(&self, alpha: &[u8], beta: &R) -> Result<(PublicKey<R>, Key, Key), VdafError> {
+        if beta.is_zero() {
+            return Err(VidpfError::InvalidParameter("beta cannot be zero".into()).into());
+        }
+
+        // Key Generation.
+        let k0 = Key::gen(ServerID::S0)?;
+        let k1 = Key::gen(ServerID::S1)?;
+
+        let mut s_i = [Seed::from(&k0), Seed::from(&k1)];
+        let mut t_i = [ControlBit::from(&k0), ControlBit::from(&k1)];
+
+        let n = 8 * alpha.len();
+        let mut cw = Vec::with_capacity(n);
+        let mut cs = Vec::with_capacity(n);
+
+        for i in 0..n {
+            let alpha_i = ControlBit::from((alpha[i / 8] >> (i % 8)) & 0x1);
+
+            let Sequence(sl_0, tl_0, sr_0, tr_0) = self.prg(&s_i[0]);
+            let Sequence(sl_1, tl_1, sr_1, tr_1) = self.prg(&s_i[1]);
+
+            let s_same_0 = &Seed::conditional_select(&sl_0, &sr_0, !alpha_i.0);
+            let s_same_1 = &Seed::conditional_select(&sl_1, &sr_1, !alpha_i.0);
+
+            let s_cw = s_same_0 ^ s_same_1;
+            let t_cw_l = tl_0 ^ tl_1 ^ alpha_i ^ ControlBit::from(1);
+            let t_cw_r = tr_0 ^ tr_1 ^ alpha_i;
+            let t_cw_diff = ControlBit::conditional_select(&t_cw_l, &t_cw_r, alpha_i.0);
+
+            let s_diff_0 = &Seed::conditional_select(&sl_0, &sr_0, alpha_i.0);
+            let s_diff_1 = &Seed::conditional_select(&sl_1, &sr_1, alpha_i.0);
+            let s_tilde_i_0 = s_diff_0 ^ (t_i[0] & s_cw.borrow()).borrow();
+            let s_tilde_i_1 = s_diff_1 ^ (t_i[1] & s_cw.borrow()).borrow();
+
+            let t_diff_0 = ControlBit::conditional_select(&tl_0, &tr_0, alpha_i.0);
+            let t_diff_1 = ControlBit::conditional_select(&tl_1, &tr_1, alpha_i.0);
+            t_i[0] = t_diff_0 ^ (t_i[0] & t_cw_diff);
+            t_i[1] = t_diff_1 ^ (t_i[1] & t_cw_diff);
+
+            let (w_i_0, w_i_1): (R, R);
```

Comment:
The compiler can infer the types of the variables declared here.

Generally it's better to avoid declaring types unless the compiler needs you to. That way, if the types change, we have a little less refactoring to do.
```suggestion
            let (w_i_0, w_i_1);
```